walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,214 @@
+   */
+  private int optionalCounter(Collection<TimeRange> bestTimes, Collection<Event> events, Collection<String> optional, TimeRange range, int minOverlaps) {
+    int counterOverlaps = 0;
+    for (String opAttendee : optional) {
+      Collection<Event> eventsWithAttendee = eventFinderWithAttendee(events, opAttendee);
+      for (Event e : eventsWithAttendee){
+        if (e.getWhen().overlaps(range)) {
+          counterOverlaps += 1;
+        }
+      }
+    }
+    if (counterOverlaps < minOverlaps) {
+      bestTimes.clear();
+      bestTimes.add(range);
+      return counterOverlaps;
+    }
+    else if (counterOverlaps == minOverlaps) {
+      bestTimes.add(range);
+    }
+    return minOverlaps;
+  }
+
+  /**
+   * Determines the TimeRanges where the mandatory and as many optional attendees as possible can attend the meeting
+   * @param events events of the day
+   * @param request includes mandatory and optional attendees, and the duration of the meeting
+   * @return collection of TimeRanges where all mandatory and the greatest number of optional attendees can be at the meeting
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long duration = request.getDuration();
+ 
+    // get mandatory and optional attendees
+    Collection<String> mandatory = request.getAttendees();
+    Collection<String> optional = request.getOptionalAttendees();
+    
+    // sort the events to be in order by start time
+    List<Event> events_list = new ArrayList<Event>(events); 
+    Collections.sort(events_list, Event.ORDER_BY_START); 
+    List<TimeRange> possibleTimes = new ArrayList<>();
+
+    // if the duration of meeting is invalid, then there are no meeting times.
+    if (duration > TimeRange.getTimeInMinutes(23, 59) || duration < 0) {
+      return possibleTimes;
+    }
+    
+    // if there are no events for the day, then the meeting can be any time.
+    if (events.size() == 0) {
+      possibleTimes.add(TimeRange.WHOLE_DAY);
+      return possibleTimes;
+    }
+    
+    // If there are optional attendees
+    if (optional.size() > 0) {
+
+      List<String> allAttendees = new ArrayList<>(mandatory);
+      allAttendees.addAll(optional);
+      // Create new meeting request with mandatory and optional attendees as mandatory
+      MeetingRequest new_request = new MeetingRequest(allAttendees, duration);
+
+      // Find possible time ranges with all mandatory and all optional attendees
+      Collection<TimeRange> slotsOptional = intervalFinder(events_list, new_request);
+      
+      // if there is a time with all mandatory and optional attendees, then return
+      if (slotsOptional.size() > 0) {
+        return slotsOptional;
+      }
+      // if there is no time where everyone can make it, and there is more than one optional person
+      else if (optional.size() > 1) {
```

Comment:
```suggestion
      } else if (optional.size() > 1) {
      // if there is no time where everyone can make it, and there is more than one optional person
```

Same comment for the other places.